### PR TITLE
Improve statistics collection from control files

### PR DESCRIPTION
**Pull request by marcusburghardt**
#### Description:

The `controleval.py` was recently introduced and it a very useful tool to process the control files.
However, there was space for improvements in order to make it more powerfull.

This PR basically cover the following aspects:
- Improved the readability of the code by using more descriptive names in functions and variables.
- Refactored the statistics calculations
- Reviewed and simplified the statistics calculations
- Included controls to validate arguments passed by the user
- Removed unnecessary/duplicated code
- Included options to make the output much more flexible
- Improve the statistics output by formatting the statistics output in the same standard.

This PR also incremented some statistics in the JSON output but the change was very conservative.
Since the statistics calculations were refactored, I indent to change the json conversion in order to use the same functions and this way creating a single source of data. However, I don't intend to do this in the same PR. It will be done after this PR is merged.

#### Rationale:

The control files are getting more and more important and it is critical to have good tools to analyse them.

#### Review Hints:

This PR tried to improve or extend the existing functionalities without inclusion of new features.
New features will be included in other PRs.
All the changes are separated in relevant commits, which contains a short description of the changes and the reasons behind them.
Therefore, I recommend to review this PR commit by commit, chronologically.

These are the nice command examples to see the changes:
```
$ utils/controleval.py stats -i cis_rhel8 -l l1_server -p rhel8
$ utils/controleval.py stats -i cis_rhel8 -l l1_server -p rhel8 --show-controls
$ utils/controleval.py stats -i cis_rhel8 -l l1_server -p rhel8 --show-controls -s pending
$ utils/controleval.py stats -i cis_rhel8 -l l1_server -p rhel8 -o json | jq
```

### CODE REVIEW

File: utils/controleval.py [hunk 1]
```diff
@@ -49,95 +49,160 @@ def validate_args(ctrlmgr, args):
...
+                                    - control_list[controls.Status.NOT_APPLICABLE]\
+                                    - control_list[controls.Status.PARTIAL]\
+                                    - control_list[controls.Status.PENDING]\
+                                    - control_list[controls.Status.PLANNED]\
+                                    - control_list[controls.Status.SUPPORTED]
+    control_list['partial coverage'] = ctrls - control_list[controls.Status.DOES_NOT_MEET]\
+                                    - control_list[controls.Status.NOT_APPLICABLE]\
+                                    - control_list[controls.Status.PENDING]\
+                                    - control_list[controls.Status.PLANNED]\
+                                    - control_list[controls.Status.SUPPORTED]
```
Comment (jan-cerny) [inline utils/controleval.py:125]: Could the common items be extracted out to a variable?

File: utils/controleval.py [hunk 2]
```diff
@@ -162,22 +231,27 @@ def print_to_json(ctrls, product, id, level):
     data["addressed_controls"]["all"] = get_id_array(ctrls)
     data["addressed_controls"]["applicable"] = get_id_array(applicablelist)
     data["addressed_controls"]["assessed"] = get_id_array(assessedlist)
+    data["addressed_controls"]["notassessed"] = get_id_array(applicablelist - assessedlist)
+    data["addressed_controls"]["automated"] = get_id_array(
+        ctrllist[controls.Status.AUTOMATED])
+    data["addressed_controls"]["doesnotmeet"] = get_id_array(
+        ctrllist[controls.Status.DOES_NOT_MEET])
+    data["addressed_controls"]["documentation"] = get_id_array(
```
Comment (jan-cerny) [inline utils/controleval.py:239]: Think if it would be better to transofrm this to a loop.

File: utils/controleval.py [hunk 3]
```diff
@@ -49,95 +49,160 @@ def validate_args(ctrlmgr, args):
...
+    if status not in status_count:
+        print("Error: The informed status is not available")
+        print_options(status_count)
+        exit(1)
+
+    if status_count[status] > 0:
+        print("\nList of the {status} ({total}) controls:".format(total=status_count[status],
+                                                                status=status))
+        for ctrl in control_list[status]:
+            print("{id:>16} - {title}".format(id=ctrl.id, title=ctrl.title))
```
Comment (jan-cerny) [inline utils/controleval.py:166]: It would be nice to sort the controls according to their IDs.

**Review by jan-cerny — COMMENTED** [review]
amazing

File: utils/controleval.py (hunk 1 above)
Comment (marcusburghardt) [reply utils/controleval.py:125]: Good idea. This can indeed be improved. Thanks

File: utils/controleval.py (hunk 2 above)
Comment (marcusburghardt) [reply utils/controleval.py:239]: Hum, interesting. I will send another commit to fix the CodeClimate founds and will already see how to make this function even better. Thanks

File: utils/controleval.py (hunk 3 above)
Comment (marcusburghardt) [reply utils/controleval.py:166]: Yes. That would be great and I even considered to do that. However, they are currently in a set, so I couldn't easily order them. But this is in my todo list to go deeper and improve in another PR.

File: utils/controleval.py (hunk 2 above)
Comment (marcusburghardt) [reply utils/controleval.py:239]: Actually, as informed in the PR description, I also want to refactor this function but I was conservative with it in this PR. I will refactor it in a different PR so I the authors can also review it separated. The idea is to use the new functions and apply the same logic used to generate the "human" output to the json output.

File: utils/controleval.py (hunk 1 above)
Comment (marcusburghardt) [reply utils/controleval.py]: Done.

**Comment by marcusburghardt** [comment]
Regarding the CodeClimate report, I am aware of the issue related to the `print_to_json` function. The function is indeed too long and definitely should be refactored. However, to do it in the best way, some keys names should be adjusted. For this reason, I will treat this function individually in a separate PR where the function authors can also validate the json output after the changes.

**Comment by marcusburghardt** [comment]
I don't know why CodeClimate is complaining about the line [276](https://codeclimate.com/github/ComplianceAsCode/content/pull/9660#issues). I can't reproduce the issue locally. Any idea?

**Comment by jan-cerny** [comment]
> I don't know why CodeClimate is complaining about the line [276](https://codeclimate.com/github/ComplianceAsCode/content/pull/9660#issues). I can't reproduce the issue locally. Any idea?

It somehow doesn't like the vertical indentation in Python. I think you would be safe with the hanging indentation, aligned on multiples of 4. Don't put the backslash (`\`) at the end of the line, instead put the opening parenthesis as the last character on the line. Then, inside parenthesis you can have multiple string literals in double quotes and they automatically join toghether. Something like:

```
    parser.add_argument(
        '--controls-dir', default='./controls/', help=(
            "Directory that contains control files with policy controls. "
            "e.g.: ~/scap-security-guide/controls"))
    subparsers = parser.add_subparsers(dest='subcmd', required=True)
```

**Comment by jan-cerny** [comment]
@marcusburghardt It's because of editorconfig set to multiples of 4 probably takes precendence over PEP8 in Code Climate. Take a look to https://github.com/ComplianceAsCode/content/pull/9653

**Review by jan-cerny — APPROVED** [review]

**Pull request merged**